tests/test_experiment.py

```diff
@@ -389,3 +393,135 @@ def impl(*a, **kw):
     assert len(calls) == 1, "Launcher's `get_status` was called more than once"
     (call,) = calls
     assert call == ((id_,), {}), "IDs were not de-duplicated"
+
+
+def test_wait_handles_empty_call_args(experiment):
+    """An exception is raised when asked to wait for the completion of no jobs"""
+    with pytest.raises(ValueError, match="No job ids"):
+        experiment.wait()
+
+
+def test_wait_does_not_block_unknown_id(experiment):
+    """If an experiment does not recognize a job id, it should not wait for its
+    completion
+    """
+    now = time.perf_counter()
+    experiment.wait(dispatch.create_job_id())
+    assert time.perf_counter() - now < 1
+
+
+def test_wait_calls_prefered_impl(make_populated_experment, monkeypatch):
+    """Make wait is calling the expected method for checking job statuses.
+    Right now we only have the "polling" impl, but in future this might change
+    to an event based system.
+    """
+    exp = make_populated_experment(1)
+    ((_, (id_, *_)),) = exp._launch_history.group_by_launcher().items()
+    was_called = False
+
+    def mocked_impl(*args, **kwargs):
+        nonlocal was_called
+        was_called = True
+
+    monkeypatch.setattr(exp, "_poll_for_statuses", mocked_impl)
+    exp.wait(id_)
+    assert was_called
+
+
+@pytest.mark.parametrize(
+    "num_polls",
+    [
+        pytest.param(i, id=f"Poll for status {i} times")
+        for i in (1, 5, 10, 20, 100, 1_000)
+    ],
+)
+@pytest.mark.parametrize("verbose", [True, False])
+def test_poll_status_blocks_until_job_is_completed(
+    monkeypatch, make_populated_experment, num_polls, verbose
+):
+    """Make sure that the polling based impl blocks the calling thread. Use
+    varying number of polls to simulate varying lengths of job time for a job
+    to complete.
+
+    Additionally check to make sure that the expected log messages are present
+    """
+    exp = make_populated_experment(1)
```

Comment:
typo: change to "make_populated_experiment"

Reply:
It's hard to believe this is my native language isn't it, hahahaha